src/Eshopworld.WorkerProcess/CosmosDistributedLock/CosmosDistributedLockStore.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Net;
+using System.Threading.Tasks;
+using Eshopworld.Core;
+using EShopworld.WorkerProcess.Configuration;
+using EShopworld.WorkerProcess.Telemetry;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Extensions.Options;
+using Polly;
+using Polly.Retry;
+
+namespace EShopworld.WorkerProcess.CosmosDistributedLock
+{
+    public class CosmosDistributedLockStore : ICosmosDistributedLockStore
+    {
+        private readonly IDocumentClient _documentClient;
+        private readonly IOptions<CosmosDataStoreOptions> _options;
+        private readonly AsyncRetryPolicy _retryPolicy;
+        private readonly IBigBrother _telemetry;
+
+        public CosmosDistributedLockStore(IDocumentClient documentClient, IOptions<CosmosDataStoreOptions> options, IBigBrother telemetry)
+        {
+            _documentClient = documentClient;
+            _options = options;
+            _telemetry = telemetry;
+            _retryPolicy = CreateRetryPolicy();
+        }
+
+        public async Task InitialiseAsync()
+        {
+            await _documentClient.CreateDatabaseIfNotExistsAsync(new Database { Id = _options.Value.Database })
+                .ConfigureAwait(false);
+
+            await CreateLockClaimsCollection();
+        }
+
+        private async Task CreateLockClaimsCollection()
+        {
+            var collectionDefinition = new DocumentCollection
+            {
+                Id = _options.Value.DistributedLocksCollection
+            };
+
+            await _documentClient.CreateDocumentCollectionIfNotExistsAsync(
+                UriFactory.CreateDatabaseUri(_options.Value.Database),
+                collectionDefinition,
+                new RequestOptions { OfferThroughput = _options.Value.OfferThroughput }).ConfigureAwait(false);
+        }
+
+        public async Task<bool> TryClaimLockAsync(IDistributedLockClaim claim)
+        {
+            return await _retryPolicy.ExecuteAsync(async () =>
+            {
+                try
+                {
+                    var response = await _documentClient.CreateDocumentAsync(
+                        UriFactory.CreateDocumentCollectionUri(_options.Value.Database, _options.Value.DistributedLocksCollection),
+                        claim,
+                        new RequestOptions
+                        {
+                            ConsistencyLevel = _options.Value.ConsistencyLevel
+                        }).ConfigureAwait(false);
+
+                    return response.StatusCode == HttpStatusCode.Created;
+                }
+                catch (DocumentClientException ex) when (ex.StatusCode == HttpStatusCode.Conflict)
+                {
+                    // document was created before, lock acquisition failed
```

Comment:
Would be nice to built-in some default retry/wait mechanism (here or in the higher layer) so consumers wait pre-defined window of time and if lock still not acquired, then we get an exception

Reply:
happy to do it in separate PR, though

Reply:
Can you create a regular story in the backlog for that, please?